jctools-core/src/main/java/org/jctools/queues/MpscArrayQueue.java

```diff
@@ -331,32 +338,78 @@ public E poll()
         // Copy field to avoid re-reading after volatile load
         final E[] buffer = this.buffer;
 
-        // If we can't see the next available element we can't poll
+        long seenPIndex = seenProducerIndex;
         E e = lvElement(buffer, offset); // LoadLoad
-        if (null == e)
+        if (e != null)
         {
-            /*
-             * NOTE: Queue may not actually be empty in the case of a producer (P1) being interrupted after
-             * winning the CAS on offer but before storing the element in the queue. Other producers may go on
-             * to fill up the queue after this element.
-             */
-            if (cIndex != lvProducerIndex())
+            incrementConsumerIndex(buffer, offset, cIndex, seenPIndex);
+            return e;
+        }
+        if (seenPIndex <= cIndex && (seenPIndex = lvProducerIndex()) == cIndex)
+        {
+            return null;
+        }
+        return pollMissPath(buffer, offset, cIndex, seenPIndex);
+    }
+
+    private void incrementConsumerIndex(final E[] buffer, long offset, long cIndex, final long pIndex) {
+        // This is called after the current cIndex slot has been read, and moves the consumer index
+        // to the next non-CONSUMED slot, nulling all preceding slots
+        spElement(buffer, offset, null);
+        // If no lookahead has happened then pIndex <= cIndex + 1 and we won't enter the loop
+        while (++cIndex < pIndex && lpElement(buffer, offset = calcElementOffset(cIndex)) == CONSUMED)
```

Comment:
Try to turn this long based loop into an int based one or use batch nulling (Arrays::fill? probably is worse) in 2 steps, by checking array wrapping, inf needeed. The former should be a better option

Reply:
Arrays::fill optimization is only going to be a win if this is a very big gap, which is very unlikely.